src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/SecurityServiceDispatcher.cs

```diff
@@ -0,0 +1,593 @@
+                FaultReason faultReason = new FaultReason(SR.Format(SR.SecureConversationCancelNotAllowedFaultReason), System.Globalization.CultureInfo.InvariantCulture);
+                secureConversationCloseNotSupportedFault = MessageFault.CreateFault(faultCode, faultReason);
+            }
+            return secureConversationCloseNotSupportedFault;
+        }
+
+        void ThrowIfSecureConversationCloseMessage(Message message)
+        {
+            if (message.Headers.Action == this.secureConversationCloseAction)
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperWarning(new MessageSecurityException(SR.Format(SR.SecureConversationCancelNotAllowedFaultReason), null, GetSecureConversationCloseNotSupportedFault()));
+            }
+        }
+       
+        [SecuritySafeCritical]
+        internal SecurityProtocolCorrelationState VerifyIncomingMessage(ref Message message, TimeSpan timeout, params SecurityProtocolCorrelationState[] correlationState)
+        {
+            if (message == null)
+            {
+                return null;
+            }
+            if (this.SecurityProtocol == null)
+                throw new Exception("SecurityProtocol is null");
+            ThrowIfSecureConversationCloseMessage(message);
+           // using (this.ApplyHostingIntegrationContext(message))
+           // {
+                return this.SecurityProtocol.VerifyIncomingMessage(ref message, timeout, correlationState);
+           // }
+        }
+
+
+        [SecuritySafeCritical]
+        internal void VerifyIncomingMessage(ref Message message, TimeSpan timeout)
+        {
+            if (message == null)
+            {
+                return;
+            }
+            ThrowIfSecureConversationCloseMessage(message);
+           // using (this.ApplyHostingIntegrationContext(message))
+           // {
+                this.SecurityProtocol.VerifyIncomingMessage(ref message, timeout);
+           // }
+        }
+    }
+    class SecurityReplyChannel : ServerSecurityChannel<IReplyChannel>, IReplyChannel
+    {
+        bool sendUnsecuredFaults;
+        internal static readonly SecurityStandardsManager defaultStandardsManager = SecurityStandardsManager.DefaultInstance;
+
+        public SecurityReplyChannel(SecurityServiceDispatcher securityDispatcher, IReplyChannel outerChannel)
+            : base(outerChannel)
+        {
+            this.SecurityServiceDispatcher = securityDispatcher;
+        }
+
+        public EndpointAddress LocalAddress => throw new NotImplementedException();
+
+        public IServiceChannelDispatcher ChannelDispatcher { get; set; }
+
+        public SecurityServiceDispatcher SecurityServiceDispatcher { get; }
+        public CommunicationState State => this.OuterChannel.State;
+       
+        public virtual XmlDictionaryString RenewAction
+        {
```

Comment:
This doesn't belong here. This should be getting pulled from SecuritySessionSecurityTokenAuthenticator. Same with IssueAction. You have hard coded this to a single version of SecureConversation. There are other types of token that can be requested beyond SecureConversation so this is a massive limitation you've added.